solidity/test/bridge/Bridge.test.ts

```diff
@@ -2233,229 +2252,1840 @@ describe("Bridge", () => {
+
+                        let outcome: Promise<RedemptionScenarioOutcome>
+
+                        before(async () => {
+                          await createSnapshot()
+
+                          outcome = runRedemptionScenario(data)
+                        })
+
+                        after(async () => {
+                          await restoreSnapshot()
+                        })
+
+                        // Should be deemed as valid change though rejected
+                        // because this change is a single output.
+                        it("should revert", async () => {
+                          await expect(outcome).to.be.revertedWith(
+                            "Redemption transaction must process at least one redemption"
+                          )
+                        })
+                      }
+                    )
+
+                    context(
+                      "when the single output is a legal P2WPKH change with a non-zero value",
+                      () => {
+                        const data: RedemptionTestData = SingleP2WPKHChange
+
+                        let outcome: Promise<RedemptionScenarioOutcome>
+
+                        before(async () => {
+                          await createSnapshot()
+
+                          outcome = runRedemptionScenario(data)
+                        })
+
+                        after(async () => {
+                          await restoreSnapshot()
+                        })
+
+                        // Should be deemed as valid change though rejected
+                        // because this change is a single output.
+                        it("should revert", async () => {
+                          await expect(outcome).to.be.revertedWith(
+                            "Redemption transaction must process at least one redemption"
+                          )
+                        })
+                      }
+                    )
+
+                    context(
+                      "when the single output is an illegal P2SH change with a non-zero value",
+                      () => {
+                        const data: RedemptionTestData = SingleP2SHChange
+
+                        let outcome: Promise<RedemptionScenarioOutcome>
+
+                        before(async () => {
+                          await createSnapshot()
+
+                          outcome = runRedemptionScenario(data)
+                        })
+
+                        after(async () => {
+                          await restoreSnapshot()
+                        })
+
+                        // We have this case because P2SH script has a 20-byte
+                        // payload which may match the 20-byte wallet public
+                        // key hash though it should be always rejected as
+                        // non-requested output. There is no need to check for
+                        // P2WSH since the payload is always 32-byte there.
```

Comment:
Yep. We only need to bother about 20-bytes hashes. This is because the wallet public key hash has always 20 bytes and we must make sure no redemption request uses it as a redeemer script to not confuse an output that will try to handle that request with a proper change output also referencing the wallet public key hash.

Reply:
Can you add your response to my question to the end of the comment for this test? This clarifies a lot.

Reply:
Sure!

Reply:
See https://github.com/keep-network/tbtc-v2/pull/155/commits/d83e9046d941b5fd2d31d9d688ccb0c4f04ea0c3